stage0/src/acpi.rs

```diff
@@ -0,0 +1,411 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// These data structures (and constants) are derived from qemu/hw/acpi/bios-linker-loader.c that
+// defines the interface.
+
+use crate::fw_cfg::FwCfg;
+use core::{
+    ffi::CStr,
+    fmt::{Debug, Formatter, Result as FmtResult},
+    mem::{zeroed, MaybeUninit},
+};
+use strum::FromRepr;
+
+// RSDP has to be within the first 1K of EBDA, so we treat it separately.
+// See Section 5.2.5 of the ACPI Specification, Release 6.5 for more details about RSDP, it's size,
+// structure, and where it should be placed.
+const RSDP_SIZE: usize = 36;
+// Rest of the EBDA. The full size of EBDA is 128K, but let's reserve the whole 1K for the RSDP.
+const EBDA_SIZE: usize = 127 * 1024;
+#[link_section = ".ebda.rsdp"]
+static mut RSDP: MaybeUninit<[u8; RSDP_SIZE]> = MaybeUninit::uninit();
+#[link_section = ".ebda"]
+static mut EBDA: MaybeUninit<[u8; EBDA_SIZE]> = MaybeUninit::uninit();
+
+// Safety: we include a nul byte at the end of the string, and that is the only nul byte.
+const TABLE_LOADER_FILE_NAME: &CStr =
+    unsafe { CStr::from_bytes_with_nul_unchecked(b"etc/table-loader\0") };
+const RSDP_FILE_NAME: &CStr = unsafe { CStr::from_bytes_with_nul_unchecked(b"etc/acpi/rsdp\0") };
```

Comment:
Is this running inside Restricted Kernel ?
How will we access files without having a file system?

Reply:
Worse, this is stage0, which is the code running _before_ we get to Restricted Kernel :)

The QEMU `fw_cfg` exposes a really crude file system-like API. It'd be better to think about it as a sort of really low-level key-value interface, and these are the keys. QEMU just calls them "files", so we use the same terminology throughout.